salalib/axialmap.cpp

```diff
@@ -425,75 +425,116 @@ void ShapeGraph::unlinkFromShapeMap(const ShapeMap& shapemap)
 
 // Method 1: direct linkage of endpoints where they touch
 
-void ShapeGraph::makeNewSegMap()
-{
-   // now make a connection set from the ends of lines:
-   std::vector<Connector> connectionset;
-   std::map<int,Line> lineset;
-   for (auto shape: m_shapes) {
-      if (shape.second.isLine()) {
-         connectionset.push_back(Connector());
-         lineset[shape.first] = shape.second.getLine();
-      }
-   }
+void ShapeGraph::makeNewSegMap(Communicator *comm) {
+    // now make a connection set from the ends of lines:
+    struct LineConnector {
+        const Line &m_line;
+        Connector &m_connector;
+        int m_index;
+        LineConnector(const Line &line, Connector &connector, int index)
+            : m_line(line), m_connector(connector), m_index(index) {}
+    };
+
+    std::vector<Connector> connectionset;
+    for (auto &shape : m_shapes) {
+        if (shape.second.isLine()) {
+            connectionset.emplace_back();
+        }
+    }
+    std::map<size_t, LineConnector> lineConnectors;
+    auto connectionIter = connectionset.begin();
+    int connectionIdx = 0;
+    for (auto &shape : m_shapes) {
+        if (shape.second.isLine()) {
+            lineConnectors.insert(std::make_pair(
+                shape.first, LineConnector(shape.second.getLine(), *connectionIter, connectionIdx)));
+            connectionIter++;
+            connectionIdx++;
+        }
+    }

```

Comment:
These two loops should still be one loop - you never look past the current connector in the connectionset, so no need to prepopulate, and then iterate it.

Reply:
My understanding is that emplace_back() may re-allocate elements and thus invalidate references (which is what the LineConnector keeps) in the process of growing the container. I tried it anyway and that seems to be the problem and that's also potentially why I did it like that in the first place. A potential solution would be to reserve() the size of the container, however we don't know how many of the shapes are lines meaning we'd still need another loop. Happy to take suggestions/corrections!

Reply:
How about replacing `connectionIter++;` with (should be pre-increment anyway...) with `connectionIter = connectionset.end() - 1;` after emplacing back the current element, but before creating the connector?

Reply:
Or even better - create the connector with `connectionset.back()`.

Reply:
Use a `std::deque` instead of a `std::vector`: https://en.cppreference.com/w/cpp/container/deque
When expanding, it creates a new array and links that rather than moving existing elements. It does that for other reasons, but the result would save you here.